From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: UN1CA <noreply@unica.local>
Date: Sun, 21 Jun 2026 00:00:00 +0000
Subject: [PATCH] Allow local SCPM allowlist packages

Samsung SCPM can overwrite AudioMirroring's private ALLOW_LIST with only
com.spotify.music. Keep the stock SCPM flow intact, but also accept exact
packages from Settings.System.unica_scpm_allowlist_packages at runtime.
---
 .../service/MediaRoute2MirroringProviderService.smali          |  2 ++
 smali/k1/b.smali                                               | 53 +++++++++++++++++
 2 files changed, 55 insertions(+)

diff --git a/smali/com/samsung/android/audiomirroring/service/MediaRoute2MirroringProviderService.smali b/smali/com/samsung/android/audiomirroring/service/MediaRoute2MirroringProviderService.smali
index 1111111..2222222 100644
--- a/smali/com/samsung/android/audiomirroring/service/MediaRoute2MirroringProviderService.smali
+++ b/smali/com/samsung/android/audiomirroring/service/MediaRoute2MirroringProviderService.smali
@@ -2445,6 +2445,8 @@

     iput-object v1, p0, Lcom/samsung/android/audiomirroring/service/MediaRoute2MirroringProviderService;->g:Landroid/content/Context;

+    sput-object v1, Lk1/b;->h:Landroid/content/Context;
+
     invoke-static {}, LW/a;->m()Ljava/lang/String;

     move-result-object v1
diff --git a/smali/k1/b.smali b/smali/k1/b.smali
index 3333333..4444444 100644
--- a/smali/k1/b.smali
+++ b/smali/k1/b.smali
@@ -19,6 +19,8 @@

 .field public static g:I

+.field public static h:Landroid/content/Context;
+

 # direct methods
 .method static constructor <clinit>()V
@@ -61,6 +63,57 @@

     invoke-static {p0, v0}, LE1/k;->e(Ljava/lang/Object;Ljava/lang/String;)V

+    sget-object v0, Lk1/b;->h:Landroid/content/Context;
+
+    if-eqz v0, :cond_unica_settings_done
+
+    invoke-virtual {v0}, Landroid/content/Context;->getContentResolver()Landroid/content/ContentResolver;
+
+    move-result-object v0
+
+    const-string v1, "unica_scpm_allowlist_packages"
+
+    invoke-static {v0, v1}, Landroid/provider/Settings$System;->getString(Landroid/content/ContentResolver;Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_unica_settings_done
+
+    const-string v1, "\\n"
+
+    invoke-virtual {v0, v1}, Ljava/lang/String;->split(Ljava/lang/String;)[Ljava/lang/String;
+
+    move-result-object v0
+
+    const/4 v1, 0x0
+
+    :goto_unica_settings_loop
+    array-length v2, v0
+
+    if-ge v1, v2, :cond_unica_settings_done
+
+    aget-object v2, v0, v1
+
+    invoke-virtual {v2}, Ljava/lang/String;->trim()Ljava/lang/String;
+
+    move-result-object v2
+
+    invoke-static {v2, p0}, LE1/k;->a(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v2
+
+    if-eqz v2, :cond_unica_settings_next
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_unica_settings_next
+    add-int/lit8 v1, v1, 0x1
+
+    goto :goto_unica_settings_loop
+
+    :cond_unica_settings_done
     sget-object v0, Lk1/b;->d:Ljava/util/ArrayList;

     invoke-interface {v0}, Ljava/util/List;->size()I
--
2.43.0
